### DISCUSSION TITLE
To make sendmsg zerocopy more flexable

### DISCUSSION THREAD

**Discussion by pyhd** [axboe/liburing#1052] [Ideas]
According to the doc:
> Copy avoidance is not a free lunch. As implemented, with page pinning, it replaces per byte copy cost with page accounting and completion notification overhead. As a result, MSG_ZEROCOPY is generally only effective at writes over around 10 KB.

Could it be possible to add a flag or op code with a new parameter(threshold), just to tell sendmsg when it is fit to use copy or zero_copy method? Given notification overhead is avoided by io_uring, the optimal vaule shall be less than 10KB, but page accounting is still not free.

**Comment by pyhd** [comment]
Fine, I found the answer #740. `io_uring_prep_send_zc_fixed` saves the cycles of page accounting, but for the sake of UDP GSO, I have to deal with sendmsg. What a pity!

**Comment by isilence** [comment]
> Could it be possible to add a flag or op code with a new parameter(threshold), just to tell sendmsg when it is fit to use copy or zero_copy method? Given notification overhead is avoided by io_uring, the optimal vaule shall be less than 10KB, but page accounting is still not free.

If the threshold is passed from the userspace, userspace may as easily choose b/w `IORING_OP_SENDMSG` and `IORING_OP_SENDMSG_ZC`.

```C
if (size > 10K)
    io_uring_prep_sendmsg_zc();
else
    io_uring_prep_sendmsg();
```

Maybe I didn't get the idea?

And I need to note that the kernel itself doesn't know what the threshold should be, it depends on how apps use io_uring, what is completion batching and the general pattern, on the CPU and the cost of the send path and so on. 

> Fine, I found the answer https://github.com/axboe/liburing/issues/740. io_uring_prep_send_zc_fixed saves the cycles of page accounting, but for the sake of UDP GSO, I have to deal with sendmsg. What a pity!

Right, I need to add it to `sendmsg`, but to be fair the notification overhead is still there, which is not less important, so not a free lunch anyway.
